assembly/src/compiled_library.rs

```diff
@@ -0,0 +1,184 @@
+use alloc::{collections::BTreeMap, vec::Vec};
+use vm_core::{
+    crypto::hash::RpoDigest,
+    mast::{MastForest, MerkleTreeNode},
+};
+
+use crate::{
+    ast::{FullyQualifiedProcedureName, ProcedureIndex, ProcedureName},
+    CompiledLibraryError, LibraryNamespace, LibraryPath, Version,
+};
+
+/// A procedure's name, along with its module path.
+///
+/// The only difference between this type and [`FullyQualifiedProcedureName`] is that
+/// [`CompiledFullyQualifiedProcedureName`] doesn't have a [`crate::SourceSpan`].
+pub struct CompiledFullyQualifiedProcedureName {
+    /// The module path for this procedure.
+    pub module_path: LibraryPath,
+    /// The name of the procedure.
+    pub name: ProcedureName,
+}
+
+impl CompiledFullyQualifiedProcedureName {
+    pub fn new(module_path: LibraryPath, name: ProcedureName) -> Self {
+        Self { module_path, name }
+    }
+}
+
+impl From<FullyQualifiedProcedureName> for CompiledFullyQualifiedProcedureName {
+    fn from(fqdn: FullyQualifiedProcedureName) -> Self {
+        Self {
+            module_path: fqdn.module,
+            name: fqdn.name,
+        }
+    }
+}
+
+/// Stores the name and digest of a procedure.
+#[derive(Debug, Clone)]
+pub struct ProcedureInfo {
+    pub name: ProcedureName,
+    pub digest: RpoDigest,
+}
+
+/// Represents a library where all modules modules were compiled into a [`MastForest`].
+pub struct CompiledLibrary {
```

Comment:
I think we should call this just `Library`, or maybe `MastLibrary`. There isn't a non-compiled `Library` anymore (since this replaces it). I think it also makes sense to move all this into the `library` namespace, and strip out the stuff in there that we no longer care about.